### PR TITLE
Integration tests for get certificate usage api

### PR DESCRIPTION
**Pull request by IsuruMaduranga**
Resolves https://github.com/wso2/api-manager/issues/499

### CODE REVIEW

File: modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java [hunk 1]
```diff
@@ -325,6 +326,145 @@ public okhttp3.Call deleteEndpointCertificateByAliasAsync(String alias, final Ap
...
+        <tr><td> 400 </td><td> Bad Request. Invalid request or validation error. </td><td>  -  </td></tr>
+        <tr><td> 404 </td><td> Not Found. The specified resource does not exist. </td><td>  -  </td></tr>
+        <tr><td> 500 </td><td> Internal Server Error. </td><td>  -  </td></tr>
+     </table>
+     */
+    public okhttp3.Call getCertificateUsageByAliasCall(String alias, Integer limit, Integer offset, final ApiCallback _callback) throws ApiException {
+        Object localVarPostBody = null;
+
+        // create path and map variables
+        String localVarPath = "/endpoint-certificates/{alias}/usage"
```
Comment (dnwick) [inline modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java:350]: Add a suitable constant

File: modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java [hunk 2]
```diff
@@ -325,6 +326,145 @@ public okhttp3.Call deleteEndpointCertificateByAliasAsync(String alias, final Ap
...
+        Map<String, Object> localVarFormParams = new HashMap<String, Object>();
+        final String[] localVarAccepts = {
+            "application/json"
+        };
+        final String localVarAccept = localVarApiClient.selectHeaderAccept(localVarAccepts);
+        if (localVarAccept != null) {
+            localVarHeaderParams.put("Accept", localVarAccept);
+        }
+
+        final String[] localVarContentTypes = {
```
Comment (dnwick) [inline modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java:374]: why we r initializing a empty array ?

File: modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java [hunk 3]
```diff
@@ -325,6 +326,145 @@ public okhttp3.Call deleteEndpointCertificateByAliasAsync(String alias, final Ap
...
+    }
+
+    @SuppressWarnings("rawtypes")
+    private okhttp3.Call getCertificateUsageByAliasValidateBeforeCall(String alias, Integer limit, Integer offset, final ApiCallback _callback) throws ApiException {
+        
+        // verify the required parameter 'alias' is set
+        if (alias == null) {
+            throw new ApiException("Missing the required parameter 'alias' when calling getCertificateUsageByAlias(Async)");
+        }
+
```
Comment (dnwick) [inline modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java:391]: too many new lines fix formatting in all places

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java [hunk 4]
```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2022, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (dnwick) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java:2]: fix Inc

**Review by dnwick — DISMISSED** [review]

File: modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java (hunk 1 above)
Comment (IsuruMaduranga) [reply modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java:350]: This class is auto-generated

File: modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java (hunk 2 above)
Comment (IsuruMaduranga) [reply modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java:374]: This class is auto-generated

File: modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java (hunk 3 above)
Comment (IsuruMaduranga) [reply modules/integration/tests-common/clients/publisher/src/gen/java/org/wso2/am/integration/clients/publisher/api/v1/EndpointCertificatesApi.java:391]: This class is auto-generated

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java (hunk 4 above)
Comment (IsuruMaduranga) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java]: fixed

**Review by tharindu1st — APPROVED** [review]

File: modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java [hunk 5]
```diff
@@ -2535,4 +2536,19 @@ private void waitUntilStatusToBlock(String apiId, String action) throws APIManag
         }
 
     }
+
+    /**
+     * Get the Certificate usage
```
Comment (sajinieKavindya) [inline modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java:2541]: Shall we end this sentence with a dot (period)? Let's try to follow the javadoc standards as much as possible.

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+
+        this.evenAPIs = new ArrayList<>();
+        this.allAPIs = new ArrayList<>();
+
+        String providerName = user.getUserName();
+
+        Random random = new Random();
+        int limit = random.ints(20, 30).findFirst().getAsInt();
+
+        for(int i = 0; i < limit; i++) {
```
Comment (sajinieKavindya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java:88]: Let's fix the formatting here. White space is required between `for` and `(`

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+            String apiName = API_NAME_BASE + "_" + i;
+            String apiContext = API_CONTEXT_BASE + "_" + i;
+            String apiEndPointUrl = API_ENDPOINT_URL_BASE_1 + i;
+
+            if(i % 2 == 0) {
```
Comment (sajinieKavindya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java:93]: Here too

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java [hunk 8]
```diff
@@ -0,0 +1,231 @@
...
+                apiEndPointUrl = API_ENDPOINT_URL_BASE_2 + i;
+                selectedEndpointUrl = apiEndPointUrl;
```
Comment (sajinieKavindya) [inline modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java:95]: `selectedEndpointUrl` is only initialized inside this condition. Is it the correct logic here?

Comment (IsuruMaduranga) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java:95]: Yeah. This is the correct logic. We get the `apiEndPointUrl` corresponding to the last value divisible by 2 of the given range here as `selectedEndpointUrl`. There will always be a value for `selectedEndpointUrl`

File: modules/integration/tests-common/clients/publisher/src/main/resources/publisher-api.yaml [hunk 9]
```diff
@@ -11648,7 +11731,7 @@ components:
           type: string
           example: v4.1.0
         data:
-#          type: object
+          #          type: object
```
[the remark this answers is not in the document]
Comment (IsuruMaduranga) [reply modules/integration/tests-common/clients/publisher/src/main/resources/publisher-api.yaml:11734]: Fixed https://github.com/wso2/product-apim/pull/13040

File: modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java (hunk 5 above)
Comment (IsuruMaduranga) [reply modules/integration/tests-common/integration-test-utils/src/main/java/org/wso2/am/integration/test/impl/RestAPIPublisherImpl.java:2541]: Fixed https://github.com/wso2/product-apim/pull/13040

File: modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java (hunk 7 above)
Comment (IsuruMaduranga) [reply modules/integration/tests-integration/tests-backend/src/test/java/org/wso2/am/integration/tests/api/lifecycle/APIEndpointCertificateUsageTestCase.java:93]: Fixed https://github.com/wso2/product-apim/pull/13040

**Pull request merged**